index.js

```diff
@@ -877,67 +865,103 @@ class Watcher extends EventEmitter {
 
-    this.running = false
-
     this.latestDiff = 0
     this.range = range
+
+    this.current = null
+    this.previous = null
     this.stream = null
 
-    this._onappendBound = debounceify(this._onappend.bind(this))
+    this._mutex = { lock: mutexify(), release: null }
+    this._resolveOnChange = null
 
     this._opening = this._ready().catch(safetyCatch)
   }
 
   async _ready () {
     await this.bee.ready()
-    this.latestDiff = this.bee.version
+    this.current = this.bee.snapshot() // point from which to start watching
     this.opened = true
   }
 
-  async _onappend () {
-    this.running = true
+  [Symbol.asyncIterator] () {
+    return this
+  }
 
-    try {
-      await this._run()
-    } catch (err) {
-      if (!this.closed) this.emit('error', err)
-      this.destroy()
-    } finally {
-      this.running = false
-    }
+  _onappend () {
+    const resolve = this._resolveOnChange
+    this._resolveOnChange = null
+    if (resolve) resolve()
   }
 
-  async _run () {
-    if (this.closed) return
-    if (this.opened === false) await this._opening
+  async _waitForChanges () {
+    if (this.current.version < this.bee.version || this.closed) return
+
+    await new Promise(resolve => {
+      this._resolveOnChange = resolve
+    })
+  }
 
-    const snapshot = this.bee.snapshot()
-    this.stream = snapshot.createDiffStream(this.latestDiff, this.range)
+  async next () {
+    this._mutex.release = await this._mutex.lock()
 
     try {
-      for await (const data of this.stream) { // eslint-disable-line
-        this.emit('change', snapshot.version, this.latestDiff)
-        break
+      if (this.closed) return { value: undefined, done: true }
+
+      if (!this.opened) await this._opening
+
+      while (true) {
+        await this._waitForChanges()
+        if (this.closed) return { value: undefined, done: true }
+
+        if (this.previous) await this.previous.close()
+        this.previous = this.current.snapshot()
+
+        if (this.current) await this.current.close()
+        this.current = this.bee.snapshot()
+
+        this.stream = this.current.createDiffStream(this.previous.version, this.range)
+
+        try {
+          for await (const data of this.stream) { // eslint-disable-line
+            return { done: false, value: { current: this.current, previous: this.previous } }
+          }
+        } finally {
+          this.stream = null
+        }
       }
     } finally {
-      this.stream = null
-      this.latestDiff = snapshot.version
-
-      await snapshot.close()
+      this._mutex.release()
     }
   }
 
-  destroy () {
+  async destroy () {
     if (this.closed) return
     this.closed = true
 
+    this.bee._watchers.delete(this)
+
     if (this.stream && !this.stream.destroying) {
       this.stream.destroy()
     }
 
-    this.bee._watchers.delete(this)
+    this._onappend() // Continue execution being closed
+
+    if (this._mutex.lock.locked && this._mutex.release) this._mutex.release()
```

Comment:
acquire the lock and free it immediately instead, then you know the async iterator has drained plus less branching